Spigot-Server-Patches/0565-Make-Wandering-Traders-spawn-below-the-roof-if-confi.patch

```diff
@@ -0,0 +1,35 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: jmp <jasonpenilla2@me.com>
+Date: Tue, 25 Aug 2020 13:22:33 -0700
+Subject: [PATCH] Make Wandering Traders spawn below the roof if configured to
+ spawn in a dimension with a roof
+
+
+diff --git a/src/main/java/net/minecraft/server/MobSpawnerTrader.java b/src/main/java/net/minecraft/server/MobSpawnerTrader.java
+index 502cb0ea40b5df1962c5a6d4b57d34de9d413e7f..a9e1a41dde3229c0a4be885d231ca890a42041ec 100644
+--- a/src/main/java/net/minecraft/server/MobSpawnerTrader.java
++++ b/src/main/java/net/minecraft/server/MobSpawnerTrader.java
+@@ -122,10 +122,20 @@ public class MobSpawnerTrader implements MobSpawner {
+             int k = blockposition.getX() + this.a.nextInt(i * 2) - i;
+             int l = blockposition.getZ() + this.a.nextInt(i * 2) - i;
+             int i1 = iworldreader.a(HeightMap.Type.WORLD_SURFACE, k, l);
+-            BlockPosition blockposition2 = new BlockPosition(k, i1, l);
++            // Paper start - allow traders to spawn below the nether roof
++            BlockPosition.MutableBlockPosition blockposition2 = new BlockPosition.MutableBlockPosition(k, i1, l);
++            if (iworldreader.getDimensionManager().hasCeiling()) {
++                do {
++                    blockposition2.c(EnumDirection.DOWN);
++                } while (!iworldreader.getType(blockposition2).isAir());
++                do {
++                    blockposition2.c(EnumDirection.DOWN);
++                } while (iworldreader.getType(blockposition2).isAir() && blockposition2.getY() > 0);
++            }
+ 
+-            if (SpawnerCreature.a(EntityPositionTypes.Surface.ON_GROUND, iworldreader, blockposition2, EntityTypes.WANDERING_TRADER)) {
+-                blockposition1 = blockposition2;
++            if (SpawnerCreature.a(EntityPositionTypes.Surface.ON_GROUND, iworldreader, blockposition2.immutableCopy(), EntityTypes.WANDERING_TRADER)) {
++                blockposition1 = blockposition2.immutableCopy();
```

Comment:
Do we need to do an immutable copy here? If so, at least put it into its own variable to avoid doing two of them.